src/server/socket.js

```diff
@@ -1,115 +1,193 @@
-            // take the most recent.
-            hostSockets[SIM_HOST] = socket;
+function handleSimHostReady() {
+    // Resolve the deferred
+    whenSimHostReady.resolve();
 
-            socket.on('exec-success', function (data) {
-                emitToHost(APP_HOST, 'exec-success', data);
-            });
-            socket.on('exec-failure', function (data) {
-                emitToHost(APP_HOST, 'exec-failure', data);
-            });
+    setupSimHostHandlers();
 
-            socket.on('plugin-message', function (data) {
-                emitToHost(APP_HOST, 'plugin-message', data);
-            });
+    whenAppHostConnected.promise
+        .then(onAppHostConnected);
+}
 
-            socket.on('plugin-method', function (data, callback) {
-                emitToHost(APP_HOST, 'plugin-method', data, callback);
-            });
+function handleAppPluginList(data) {
+    whenSimHostReady.promise.then(function () {
+        subscribeTo(SIM_HOST, 'start', handleStart, true);
+        emitTo(SIM_HOST, 'app-plugin-list', data);
+    });
+}
 
-            socket.on('telemetry', function (data) {
-                telemetry.handleClientTelemetry(data);
-            });
+function handleStart() {
+    emitTo(APP_HOST, 'start');
+}
 
-            socket.on('debug-message', function (data) {
-                emitToHost(DEBUG_HOST, data.message, data.data);
-            });
+function setupSimHostHandlers() {
+    log.log('Setup handlers for SIM_HOST');
+
+    subscribeTo(SIM_HOST, 'exec-success', function (data) {
+        emitTo(APP_HOST, 'exec-success', data);
+    });
+
+    subscribeTo(SIM_HOST, 'exec-failure', function (data) {
+        emitTo(APP_HOST, 'exec-failure', data);
+    });
 
-            // Set up telemetry if necessary.
-            if (config.telemetry) {
-                socket.emit('init-telemetry');
+    subscribeTo(SIM_HOST, 'plugin-message', function (data) {
+        emitTo(APP_HOST, 'plugin-message', data);
+    });
+
+    subscribeTo(SIM_HOST, 'plugin-method', function (data, callback) {
+        emitTo(APP_HOST, 'plugin-method', data, callback);
+    });
+
+    subscribeTo(SIM_HOST, 'debug-message', function (data) {
+        emitTo(DEBUG_HOST, data.message, data.data);
+    });
+
+    subscribeTo(SIM_HOST, 'telemetry', function (data) {
+        telemetry.handleClientTelemetry(data);
+    });
+
+    // Set up telemetry if necessary.
+    if (config.telemetry) {
+        emitTo(SIM_HOST, 'init-telemetry');
+    }
+
+    handlePendingEmits(SIM_HOST);
+
+}
+
+function init(server) {
+    var io = require('socket.io')(server);
+
+    io.on('connection', function (socket) {
+        socket.on('register-app-host', function () {
+            log.log('APP_HOST connected to the server');
+            if (hostSockets[APP_HOST]) {
+                log.log('Overriding previously connected APP_HOST');
+                resetAppHostState();
```

Comment:
Yes. I believe I implemented the override here because while testing the refresh/reload scenario I found that we couldn't not rely on old APP_HOST `disconnect` getting before the new APP_HOST `register` message. Perhaps, it could be resolved by closing the socket or explicitly unregistering in `onbeforeunload`, but the new-registration-overrides type of policy is the safest.